Add SIR interpreter (Version 2).

ptersilie opened the pull request:
This is the second attempt at our SIR interpreter which now uses a memory map to store the locals. Before you raise your pitchforks keep two things in mind:

1) This currently doesn't support multiple frames and only interprets a single SIR body. We'll implement those together with terminators (calls etc.).
2) The offsets of the locals are currently computed at runtime, which is slow. We will soon make the necessary changes to SIR to do this for us at compile time. We can then retrieve the offsets quickly from SIR. I thought this would be better in a separate PR.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+};
+use yktrace::sir::SIR;
+
+const SUSIZE: usize = std::mem::size_of::<usize>();
```
I think it's better to just do:

```rust
use std::mem::size_of;
```

and then:

```rust
size_of::<usize>()
```

where necessary. It stops people wondering "what is `SUSIZE`?"

ptersilie replied:
This is probably me trying to be too smart. My thinking was that I don't want to have this call each time we need the size of usize. I forgot the compiler will just optimise away the call anyway.

ltratt replied:
Yeah, it's not really a call. [Rust has statically evaluated const expressions which, with my Converge CTMP hat on, look incredibly ad-hoc and verbose. But that's a separate rant.]

ptersilie replied:
Fixed in b931d23a549829cbd924ec6f1180f7cf7170bb9e.

ptersilie replied:
Fixed in b931d23.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+const SUSIZE: usize = std::mem::size_of::<usize>();
+
+pub struct SIRInterpreter {
+    map: *mut u8,
```
What are `map` and `layout`? I was expecting something like:

```rust
struct SIRInterp {
    locals: *mut u8
}
```

but maybe I've misunderstood what's needed!

ptersilie replied:
We need to hold on to the layout so we can deallocate the memory when the interpreter is dropped. Or will recreating the same layout for the `dealloc` call also work?

ltratt replied:
Once the locals are statically calculated, the `Layout` should be trivially calculable (it should become equivalent to a load of a size & alignment) and thus wouldn't need to be stored here.

ptersilie replied:
Yeah, I missed that. Will fix.

ptersilie replied:
Fixed in b5ac93e.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+pub struct SIRInterpreter {
+    map: *mut u8,
+    layout: Layout,
+    locals: Vec<usize>,
```
```suggestion
    local_offsets: Vec<usize>,
```
It wasn't immediately clear to me that these are offsets into `map`.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+        }
+
+        // Allocate memory for the locals
+        let layout = Layout::from_size_align(usize::try_from(size).unwrap(), 8).unwrap();
```
There may be a local that needs to be aligned to a bigger size. I don't know if alignment is really important for the interpreter though. If not, it should be 1 instead.

ltratt replied:
There could definitely be locals which are aligned to arbitrary sizes, so I think we need to handle that. [In practise, not many things are aligned to `>` machine word, but still...]

ptersilie replied:
Actually, I did mean `1` here. This is a typo.

ptersilie replied:
Fixed in 01b6575.

ptersilie replied:
On second thought, I changed this back to `size_of::<usize>()`, which allows us to simplify the code that reads/writes pointers, and should also give us some speedup: f5cc2d7

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+const SUSIZE: usize = std::mem::size_of::<usize>();
+
+pub struct SIRInterpreter {
+    map: *mut u8,
```
Can this be `Box<[u8]>` or `*mut [u8]`?

ptersilie replied:
What are the advantages of this?

bjorn3 replied:
The first one will avoid unsafe allocation and de-allocation code. The second one merely makes bound checking much easier.

ptersilie replied:
I think I prefer `*mut u8` over this, since we'll have to do a lot of pointer arithmetic. Though I could be convinced otherwise if there's a way to easily do that with `[u8]` as well.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /// Sets the current frame of the interpreter.
+    pub fn set_frame(&mut self, tio: usize) {
```
I'm not sure what `tio` does here, in particular why we we might be calling `to_ne_bytes` on it. Any clues?

ptersilie replied:
`tio` is a pointer to the trace inputs which I need to store inside the locals map. This will go away (hopefully) once I get rid of the pointer as usize stuff.

ptersilie replied:
Okay, this is a bit more complicated. We need `to_ne_bytes` as we are interested in the value of pointer itself, not what it points to. So we need to convert the `&IO` into a slice of bytes or a `*const u8`. If there's a better way than going through a `usize` I'm all ears.

ltratt replied:
I don't know what this function is really intending to do, so it's very hard for me to say anything sensible I'm afraid.

ptersilie replied:
At the moment, all it does is pass the pointer to the `trace inputs` to the interpreter. `Trace inputs` is what we also hand to the compiler, which contains stuff like the program counter, etc. If you look at the tests below you can see how `tio` is used.

ltratt replied:
I guess what I'm saying is: I don't understand the high-level structure that leads to this decision. My default reaction is that this code is unlikely to be correct under stack borrows, but perhaps it is, and perhaps this is the best way of doing things. I don't know.

ptersilie replied:
Since we are not writing to `tio` until `interpret` returns, this should be fine under stacked borrows. To be sure I've tested this in MIRI which seems happy: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8a3fd007a16b3db15720382f9b486611

I've added a small comment in 4f9d18c and changed the argument type from `usize` to `*mut usize` to avoid confusion in 1a5048d.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+        // counter of the interpreter.
+        let ptr = self.get_local(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            std::ptr::copy(usize::to_ne_bytes(tio).as_ptr(), ptr as *mut u8, SUSIZE);
```
Isn't `tio.to_ne_bytes()` accepable over `usize::to_ne_bytes`?

ptersilie replied:
By aligning the locals I was able to simplify this code: f5cc2d7

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+        // FIXME Later this also sets other already initialised variables as well as the program
+        // counter of the interpreter.
+        let ptr = self.get_local(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            std::ptr::copy(usize::to_ne_bytes(tio).as_ptr(), ptr as *mut u8, SUSIZE);
+        }
```
If `get_local` were to return a `&mut [u8]`, you could use `copy_from_slice` here.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+        for stmt in body.blocks[0].stmts.iter() {
+            match stmt {
+                Statement::MkRef(dest, src) => self.mkref(dest, src),
+                Statement::DynOffs { .. } => todo!("dynoffs"),
```
You can put strings in `todo` if you want, but since it prints the file and line number by default, you might as well save yourself some typing in my opinion!

ptersilie replied:
Yeah, looks like some debugging left over.

ptersilie replied:
Fixed in d48ba09.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /// Get the pointer to a Local.
+    fn get_local(&self, local: &Local) -> usize {
+        let offset = self.locals[usize::try_from(local.0).unwrap()];
+        self.map as usize + offset
+    }
```
```suggestion
    fn get_local(&self, local: Local) -> *mut u8 {
        let offset = self.locals[usize::try_from(local.0).unwrap()];
        self.map.offset(isize::try_from(offset).unwrap())
    }
```

ptersilie replied:
Fixed in 9e9ef8f.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /// Get the pointer to a Local.
+    fn get_local(&self, local: &Local) -> usize {
```
I couldn't work out why the return type was `usize` until I realised that you're trying to avoid pointers. I think this method should be called `local_ptr` and return `*mut u8`.

ptersilie replied:
Fixed in 9e9ef8f.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /// Get the pointer for an IPlace, while applying all offsets.
+    fn iplace_to_ptr(&self, place: &IPlace) -> usize {
```
Please use `*mut u8` here too.

ptersilie replied:
Fixed in 9e9ef8f.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+
+    /// Write some bytes to an IPlace. The amount of bytes is determined by the type of the
+    /// destination.
+    fn write(&mut self, dest: &IPlace, src: *const u8) {
```
```suggestion
    unsafe fn write(&mut self, dest: &IPlace, src: *const u8) {
```
Also prefer `&[u8]`. (It still has to be unsafe, as `dest` could contain a invalid or aliasing pointer.)

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,267 @@
+                ty,
+            }
+            | IPlace::Indirect { ptr: _, off: _, ty } => {
+                let size = usize::try_from(SIR.ty(ty).size()).unwrap();
```
I think the SIR should hold `usize` for this kind of things. It isn't portable between architectures anyway.

ptersilie replied:
Agreed.

ltratt commented:
At a couple of places, this PR returns pointers as `usize`. There are some use cases for that, but it's pretty dangerous in general, as it obscures the fact there's a pointer (I misread the code entirely at first), but without saving us from Rust's onerous pointer rules (see https://www.ralfj.de/blog/2018/07/24/pointers-and-bytes.html for surprising info on this). In general, pointers should be passed around as `*mut T` or `*const T` to make clear what's going on.

bjorn3 reviewed (changes requested):
There is way too much unnecessary unsafe here. Miri manages with just a few lines of unsafe code in a very hot function.

ltratt commented:
[I wonder if you're converting things to `usize` to do pointer arithmetic? If so, Rust's pointers have built-in methods to do this https://doc.rust-lang.org/std/primitive.pointer.html#method.add.]

ptersilie commented:
> [I wonder if you're converting things to `usize` to do pointer arithmetic? If so, Rust's pointers have built-in methods to do this https://doc.rust-lang.org/std/primitive.pointer.html#method.add.]

Yes, that was my intention. Missed the pointer methods. Will fix.

ltratt commented:
@ptersilie There are a lot of new commits. Do they need reviewing individually?

ptersilie commented:
After reading up on stacked borrows and playing around with MIRI I have addressed some of the comments. I've also refactored some of the code. So I think this is ready for re-review.

ptersilie commented:
> @ptersilie There are a lot of new commits. Do they need reviewing individually?

I've commented the commit hash on the individual comments above. So you only need to address the ones addressing your comment. However, the most interesting commits are the last 4, starting at 1a5048d.

ltratt commented:
I don't see any references to commits?

ptersilie commented:
> I don't see any references to commits?

https://github.com/softdevteam/yk/pull/159#discussion_r533351143
https://github.com/softdevteam/yk/pull/159#discussion_r533349940
https://github.com/softdevteam/yk/pull/159#discussion_r533349317

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -35,12 +35,13 @@ impl SIRInterpreter {
 
     /// Sets the current frame of the interpreter.
-    pub fn set_frame(&mut self, tio: usize) {
+    pub fn set_frame(&mut self, tio: *mut usize) {
```
Can you tell me what the model of the "current frame" is here, and what "tio" is? The description "sets the current frame" makes sense, but I can't relate that to the code inside this function. That's probably my bad, but I'm going to need your help to make sense of this!

ptersilie replied:
At the moment this is pretty much just a stub, since we haven't implemented the relevant part in the SIR yet, which will pass in the `tio` pointer (which points to the same trace inputs used by the yk compiler) and other initialised variables. The goal of this function is to initialise the current frame of the SIR we are going to interpret, which includes setting the program counter to the next instruction to be executed. Maybe I can add the program counter bit in this PR to make more clear what this does?

ltratt replied:
OK, so this isn't really about "setting the frame" then? `set_trace_inputs` or something?

ptersilie replied:
At the moment, no. But soon it will be. So we have two options to fix this:
1) We rename it, and name it back once it actually sets the frame.
2) I add initialisation of the program counter to this function, to make clearer that this will be setting the frame soon.

ltratt replied:
Let's name and document it such that it makes sense in this PR. Future PRs can adjust it as necessary.

ptersilie replied:
Fixed in 801159b.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -192,6 +192,8 @@ mod tests {
     fn interp(fname: &str, tio: *mut usize) {
         let body = SIR.bodies.get(fname).expect("No SIR");
         let mut si = SIRInterpreter::new(&body.local_decls);
+        // This does not alias since we won't be reading from or writing to `tio` until the
```
What's "this"?

ptersilie replied:
See if this is better: 8fe4b9e

ltratt replied:
Works for me!

ptersilie commented:
I've addressed your comments about alignments, but using Rust's API to calculate the proper alignment. Unfortunately, this meant that we need to store an extra two words for the `Layout` inside `SIRInterpreter` as we need the layout to deallocate the memory when the interpreter is dropped. At least I couldn't work out if recreating a layout using `self.offsets` would result in the same layout (my feeling is no).

ltratt commented:
> At least I couldn't work out if recreating a layout using self.offsets would result in the same layout (my feeling is no).

Why wouldn't it be deterministic? If:

1. nothing can mutate the offsets
2. you iterate over them in a deterministic order

then there shouldn't be a problem should there?

ptersilie commented:
> > At least I couldn't work out if recreating a layout using self.offsets would result in the same layout (my feeling is no).
> 
> Why wouldn't it be deterministic? If:
> 
>     1. nothing can mutate the offsets
> 
>     2. you iterate over them in a deterministic order
> 
> 
> then there shouldn't be a problem should there?

Well for a start, I don't know the size of the last element, only it's offset. So when I create a new layout using `from_size_align`, what is its size (if I take the last offset we are missing the size of the last element). And also what do I pass for `align`?

ltratt commented:
I'm confused. If you don't know that information, how can you lay anything out?

ptersilie commented:
> I'm confused. If you don't know that information, how can you lay anything out?

I know this information when I create the interpreter, because I'm passing in the locals (`local_decls`). But when we drop the interpreter we won't have access to that information any more.

ltratt commented:
Ah, hang on, we're not calculating these things statically?

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -5,36 +5,37 @@ use ykpack::{
+            let align = SIR.ty(&d.ty).align();
+            let size = SIR.ty(&d.ty).size();
+            let l = Layout::from_size_align(size.try_into().unwrap(), align.try_into().unwrap()).unwrap();
+            let (nl, s) = layout.extend(l).expect("Alignment error");
```
I think `expect` is a) almost entirely pointless and b) clutters code. Let's use `unwrap`.

ptersilie replied:
Fair.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -5,36 +5,37 @@ use ykpack::{
         // FIXME Soon this will be pre-computed and handed to us by SIR.
-        let mut size: usize = 0;
         let mut offsets = Vec::new();
+        let mut layout = Layout::from_size_align(0, 1).unwrap();
```
This feels a bit evil: would `Layout::new::<u8>()` be a bit clearer as I *think* that's what you're emulating here.

ptersilie replied:
I have no preference. I got this line from one of the examples in the docs: https://doc.rust-lang.org/std/alloc/struct.Layout.html#examples

ltratt replied:
Let's use `unwrap`.

ptersilie replied:
You mean use `Layout::new::<u8>()`?

ltratt replied:
Mea culpa, that'll teach me not to reload the GitHub tab! I see what you mean now. We can keep it as-is.

ptersilie commented:
> Ah, hang on, we're not calculating these things statically?

Not at the moment. My plan was to add that to SIR in a separate PR. However, even if SIR can give us this information statically, we still will have to store the layout. SIR can only give us this information if we tell it which SIR body we want this info for. So we either need to store a function name or the layout within the interpreter somewhere (see https://github.com/softdevteam/yk/pull/159/files#diff-11a1e7f768bc03d887b1e73424972f6621476437c93f785006b0d44582de730fR193 on how we query SIR). Once we do calls, we'll probably need a stack of layouts, so we can deallocate the memory of a SIR body when we are done with it.

ltratt commented:
I'm not sure I understood that, but let's leave that for the next PR.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -39,8 +39,8 @@ impl SIRInterpreter {
-    /// Sets the current frame of the interpreter.
-    pub fn set_frame(&mut self, tio: *mut usize) {
+    /// Inserts a pointer to the trace inputs into `locals`.
+    pub fn set_trace_inputs(&mut self, tio: *mut usize) {
```
I *think* `tio` might better be `*mut u8` to be clearer to readers that it's just a big chunk of bytes. There's not *much* difference between `*mut u8` and `*mut usize`, but the former is more conventional for "random stuff".

ptersilie replied:
I used `*mut usize` to allude to the fact, that `tio` is a pointer. It also allows me to use `write::<usize>` below, which is little nicer than `copy`. But I can also change this to `*mut u8` if you like, there's not a big difference.

ltratt replied:
`*mut T` says "this is is a pointer to zero or more `T`'s". `*mut usize` therefore says "I'm pointing to zero or more `usize`s", which isn't the case here. We really want to say "I'm pointing to zero more bytes".

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+            }
+            IPlace::Indirect { ptr, off, ty: _ty } => {
+                // Get a pointer to the Indirect.
+                let dest_ptr = self.local_ptr(&ptr.local) as *mut usize;
```
Why are we casting this `*mut usize`?

ptersilie replied:
We are doing this because `local_ptr` returns a pointer which we are going to dereference below. So casting this to `*mut usize` allows me to do:
```rust
let val = read::<usize>(ptr);
```
instead of something like
```rust
let val = 0;
copy(ptr, &val as *mut u8, size_of<usize>());
```

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+                // Get a pointer to the Indirect.
+                let dest_ptr = self.local_ptr(&ptr.local) as *mut usize;
+                // Dereference the pointer by reading its value.
+                let val = unsafe { std::ptr::read::<usize>(dest_ptr) };
```
I think this is saying that every local is a `usize`?

ptersilie replied:
In this case yes, because this is an odd situation where we have a pointer `dest_ptr` pointing to some value, that happens to also be a pointer. We know this because an `Indirect` local is always a pointer.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+            },
+            Constant::Bool(_b) => todo!(),
+            Constant::Tuple(t) => {
+                if SIR.ty(t).size() == 0 {
```
Wouldn't this be clearer as `if SIR.ty(t).size() != 0 { todo!(); }`?

ptersilie replied:
I wasn't sure if we might need to do something someday with ZSTs, but am happy to change it.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,273 @@
+    /// Creates a reference to an IPlace.
+    fn mkref(&mut self, dest: &IPlace, src: &IPlace) {
+        // Get pointer to src.
+        let src_ptr = self.iplace_to_ptr(src) as usize;
```
Do we need to use `as` here?

ptersilie replied:
Ah, got call. We can also use the `write::<usize>` here, like we do above. But I'll wait until we've discussed the other comments, because we might also want to change this to use `*mut u8` instead of `*mut usize`.

ltratt commented:
OK, I'm starting to understand more about this PR, so my last round of comments, though not hugely deep, are perhaps a bit more useful than the first round of comments!

ptersilie commented:
I've replied to your comments. See what you think.

ptersilie commented:
After some discussion offline, we decided that pointers should be cast to `*mut *mut u8`, meaning, this thing is a pointer to a pointer, pointing to some arbitrarily sized data. I've updated the code accordingly.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -99,14 +99,18 @@ impl SIRInterpreter {
+                let ptr = unsafe {
+                    // Dereference the pointer, by reading its value.
+                    let mut p = std::ptr::read::<*mut u8>(dest_ptr);
+                    // Add the offsets of the Indirect.
```
I think all of the offsets are `usize`s, right? In which case you can do `p.add(ptr.off)`. https://doc.rust-lang.org/std/primitive.pointer.html#method.add

ptersilie replied:
Unfortunately, the offsets are `i32`s, so `add` doesn't improve things.

ltratt commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,283 @@
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags & ykpack::bodyflags::TRACE_DEBUG != 0 {
```
This is out of date with the API in master. I suggest you merge master into this PR (don't rebase pretty please :p) and have a look at what's changed. The change is pretty easy.

ptersilie replied:
Fixed in 6579b5f.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,283 @@
+        }
+
+        // Allocate memory for the locals
+        let locals = unsafe { alloc(layout.pad_to_align()) };
```
```suggestion
        layout = layout.pad_to_align();
        let locals = unsafe { alloc(layout) };
```

ltratt replied:
Agreed.

ptersilie replied:
Fixed in d0d1be4.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,283 @@
+        // counter of the interpreter.
+        let ptr = self.local_ptr(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            // Write the pointer value of `tio` into locals by casting it to an integer first.
```
```suggestion
```

ltratt replied:
Agreed.

ptersilie replied:
Fixed in d0d1be4.

bjorn3 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,283 @@
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
```
```suggestion
    pub unsafe fn interpret(&mut self, body: &ykpack::Body) {
```

Given the correct `body`, this can dereference arbitrary pointers.

ptersilie replied:
Fixed in d0d1be4.

ptersilie commented:
Addressed all comments.

ltratt commented:
@vext01 Can you have a quick look at this? I don't think we want to bother Lukas with small comments at this stage, but just to check that you're happy with the overall direction.

vext01 commented on ykbh/src/lib.rs:
```diff
@@ -0,0 +1,286 @@
+        let ptr = self.local_ptr(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            // Write the pointer value of `tio` into locals.
+            std::ptr::write::<*mut u8>(ptr as *mut *mut u8, tio);
```
Why the double pointer here?

ptersilie replied:
We want to store the pointer value of `tio` and not what it points to. This avoids casting `tio` to a `usize` and makes it clearer, that `tio` is a pointer. And since we are storing a pointer `*mut u8`, the destination `ptr` needs to be a pointer to a pointer `*mut *mut u8`.

vext01 commented:
A couple of small comments.

ptersilie commented:
> A couple of small comments.

This suggest there are multiple comments, though I only saw one. Was that the only one?

vext01 commented:
My bad. Yes, please go ahead.

Next PR should do support for frames?

ptersilie commented:
> Next PR should do support for frames?

That's the plan.

ltratt commented:
Please squash.

ptersilie commented:
Squashed!

ltratt commented:
Let's update the commit message so that it gives a bit of context a) about what the SIR interpreter is b) how it works. This is an important PR and people may well look back at it, so we need to be clear with them about what's going on. Feel free to force push an update.

ptersilie commented:
I've pushed a description. It probably needs changing, so see what you think.

ltratt commented:
A quick couple of comments:

ptersilie commented:
>     1. We should explicitly say this for what RPython calls blackholing.

Hah, I had that in there at some point, but it got lost during a rewrite.

>     2. We call them "control points" not "merge points" (IIRC: @vext01 can confirm).

I think we do!

ltratt commented:
Feel free to force push an update.

ptersilie commented:
Pushed.

ltratt commented:
Happy news!

bors r+

ltratt commented:
bors r-

ltratt commented:
@ptersilie You need to rebase against master.

ptersilie commented:
Rebased!

ltratt commented:
bors r+

vext01 commented:
> realloc(): invalid old size

ooer...

ltratt commented:
@ptersilie Have you seen the CI failure?

ptersilie commented:
Yes, I'm looking into it.

ptersilie commented:
bors try

ltratt commented:
bors try

ptersilie commented:
Ok to rebase?

vext01 commented:
yup

ptersilie commented:
Rebased!

vext01 commented:
bors r+
